libherokubuildpack/src/env_layer.rs

```diff
@@ -0,0 +1,236 @@
+///                 ("MALLOC_ARENA_MAX", "2"),
+///                 ("DISABLE_SPRING", "1"),
+///             ]
+///             .into_iter(),
+///         )?;
+///         let env = layer.env.apply(Scope::Build, &env);
+///
+///#        todo!()
+///#     }
+///# }
+/// ```
+pub fn set_default<B, E, K, V>(
+    context: &BuildContext<B>,
+    layer_name: LayerName,
+    envs: E,
+) -> libcnb::Result<LayerData<GenericMetadata>, <B as Buildpack>::Error>
+where
+    B: Buildpack,
+    E: IntoIterator<Item = (K, V)> + Clone,
+    K: Into<OsString>,
+    V: Into<OsString>,
+{
+    context.handle_layer(layer_name, DefaultEnvLayer::new(envs))
+}
+
+/// Set default environment variables in a layer
+///
+/// This struct is used by the helper function `set_default`. You can also use it directly with
+/// with [`BuildContext::handle_layer`] to set default environment variables.
+pub struct DefaultEnvLayer;
+
+impl DefaultEnvLayer {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new<E, K, V, B>(env: E) -> ConfigureEnvLayer<B>
+    where
+        E: IntoIterator<Item = (K, V)> + Clone,
+        K: Into<OsString>,
+        V: Into<OsString>,
+        B: libcnb::Buildpack,
+    {
+        let mut layer_env = LayerEnv::new();
+        for (key, value) in env {
+            layer_env =
+                layer_env.chainable_insert(Scope::All, ModificationBehavior::Default, key, value);
+        }
+
+        ConfigureEnvLayer {
+            data: layer_env,
+            _buildpack: PhantomData,
+        }
+    }
+}
+
+/// Set environment variables
+///
+/// If you want to set many default environment variables you can use the
+/// `env_layer::set_default` function. If you need to set different types of environment
+/// variables you can use the `env_layer::set_envs` function. Example:
+///
+/// ```no_run
+///# use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+///# use libcnb::data::launch::{LaunchBuilder, ProcessBuilder};
+///# use libcnb::data::process_type;
+///# use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+///# use libcnb::generic::{GenericError, GenericMetadata, GenericPlatform};
+///# use libcnb::{buildpack_main, Buildpack};
+///# use libcnb::data::layer::LayerName;
+///
+///# pub(crate) struct HelloWorldBuildpack;
+///
+/// use libcnb::Env;
+/// use libcnb::data::layer_name;
+/// use libcnb::layer_env::{LayerEnv, ModificationBehavior, Scope};
+/// use libherokubuildpack::env_layer;
+///
+///# impl Buildpack for HelloWorldBuildpack {
+///#     type Platform = GenericPlatform;
+///#     type Metadata = GenericMetadata;
+///#     type Error = GenericError;
+///
+///#     fn detect(&self, _context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+///#         todo!()
+///#     }
+///
+///#     fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+///         // Don't forget to apply context.platform.env() too;
```

Comment:
This reads as if you'd always need to do that - which is not the case. If it were, libcnb should do that automatically. Can we remove this?

Reply:
I'm aiming for a higher level abstraction with this API. For someone who wants just to do one or two things and have it mostly do the right thing. I'm worried that someone like that might not understand the finer points of the difference between the different environments and when they're applied or not and just copy and paste example code.

I think most people would want the user's env vars which is confusingly labeled/buried as platform env.  One way to solve this documentation issue could be to provide a simple interface for applying platform env. Maybe something like:

```rust
let env = Env::from_current()
    .with_user_vars(context.platform.env())
```